From d7d43773667b8aa578de1574054d2acc559eabef Mon Sep 17 00:00:00 2001
From: Manoj Srivastava <srivasta@golden-gryphon.com>
Date: Sun, 27 Apr 2014 03:29:57 -0700
Subject: [PATCH 1/2] debcherry fixup patch

13a227f [master]: Moving away from git submodules, part I.
	 - no changes against upstream or conflicts
59c65d6 [master]: New upstream, updates fix for ctime.
	 - no changes against upstream or conflicts
47718f6 [mailagent]: A new version fixing mail loss.
	 - extra changes or conflicts
fd975de New debconf translations.
	 - extra changes or conflicts
13b191f Fix nroff leading comment characters.
	 - extra changes or conflicts
969da27 Fix mailagent to correctly parse top level domains
	 - extra changes or conflicts
3c008f7 tighten up declarations in order for this to compile on gcc-3.4 on amd64.
	 - extra changes or conflicts
a92866a Incorporate previous changes made for debian
	 - extra changes or conflicts
---
 Configure                 |   4 +-
 FAQ                       |  42 +++-
 README                    |  25 ++-
 agent/examples/mchk       |   2 +-
 agent/examples/profile    |   2 +-
 agent/filter/logfile.c    |   4 +-
 agent/magent.sh           |   8 +-
 agent/man/edusers.SH      |  58 +++---
 agent/man/mailagent.SH    | 498 +++++++++++++++++++++++++---------------------
 agent/man/maildist.SH     |  46 ++---
 agent/man/mailhelp.SH     |  50 ++---
 agent/man/maillist.SH     |  40 ++--
 agent/man/mailpatch.SH    |  42 ++--
 agent/man/package.SH      |  64 +++---
 agent/pl/mbox.pl          |   2 +-
 agent/pl/parse.pl         |   6 +
 agent/pl/utmp/utmp_ph.c   |   4 +-
 agent/test/TEST           |   3 +-
 agent/test/basic/config.t |   2 +-
 19 files changed, 495 insertions(+), 407 deletions(-)

diff --git a/Configure b/Configure
index 0e0efd9..63fabdd 100755
--- a/Configure
+++ b/Configure
@@ -6206,7 +6206,7 @@ $rm -f pdp11.* pdp11
 
 : determine where mail is spooled
 case "$maildir" in
-'') dflt=`./loc . /usr/spool/mail /usr/spool/mail /usr/mail /var/mail`;;
+'') dflt=`./loc . /var/spool/mail /var/spool/mail /usr/spool/mail /usr/mail /var/mail`;;
 *) dflt="$maildir";;
 esac
 echo " "
@@ -6249,7 +6249,7 @@ mailer="$ans"
 : determine where mail is spooled
 case "$mailfile" in
 '')
-	dflt=`./loc . XXX /usr/spool/mail /usr/mail /var/mail`
+	dflt=`./loc . XXX /var/spool/mail /usr/spool/mail /usr/mail /var/mail`
 	case "$dflt" in
 	XXX) dflt='%~/mailbox';;
 	*) dflt="$dflt/%L";;
diff --git a/FAQ b/FAQ
index 44d322b..c283882 100644
--- a/FAQ
+++ b/FAQ
@@ -61,13 +61,13 @@ Subject: ... 5.1 Setting Up Mailagent
 
 2b] Locate the filter program (it will be filter or filter.sh depending on
     whether you choose the C or the shell version respectively) under some
-    directory like /usr/local/lib/mailagent. From now on, we'll assume we
-    use the C filter and that it is located under /usr/local/lib/mailagent.
+    directory like /usr/lib/mailagent. From now on, we'll assume we
+    use the C filter and that it is located under /usr/lib/mailagent.
 
-3] Copy the file /usr/local/lib/mailagent/mailagent.cf as ~/.mailagent and
-   edit it to configure your system correctly. You will see two distinct
-   sections in that file and you need to set-up properly the first one, the
-   "Configuration section".
+3] Copy the file /usr/share/doc/mailagent/examples/rc/dot.mailagent.gz
+   as ~/.mailagent and edit it to configure your system correctly. You
+   will see two distinct sections in that file and you need to set-up
+   properly the first one, the "Configuration section".
 
    If you have a version of mailagent that is recent enough (at least 3.0 PL32)
    then you can create an initial configuration very easily and quickly by
@@ -117,14 +117,30 @@ Subject: ... 5.1 Setting Up Mailagent
 5b] Enter the following in ~/.rules
 
     Subject: /test/     { SAVE testing };
+    ### final default rule
+    { SAVE incoming };
 
     The meaning of that rule should be pretty obvious: If we receive a mail
     whose subject line contains the word "test", then we save that mail in
     a folder named "testing", under the default folder directory (~/Mail).
 
+    == IMPORTANT NOTE ==
+    The final default rule is needed on Debian systems since debian
+    MDA policy requires MDA's to be sgid mail, but making mailagent
+    sgid anything would be a security risk. Thus, we must *ALWAYS*
+    save any mail message that gets this far; letting it fall back to
+    the system mailbox may cause mail to be garbled.
+
+    See /usr/share/doc/mailagent/SECURITY for details
+
+    **You need that deafult rule**
+
 5c] Create a ~/.forward file as follows:
 
-    "|exec /usr/local/lib/mailagent/filter >> /export/home/ram/.bak 2>&1"
+    "|exec /usr/lib/mailagent/filter >> /export/home/ram/.bak 2>&1"
+
+	Please nore leading and  trailing double quotes are a
+	mandatory part of the line.
 
     The meaning of that line is the following: every mail should be piped
     (hence the leading "|" character) onto the filter program, and any
@@ -138,7 +154,7 @@ Subject: ... 5.1 Setting Up Mailagent
       That's a part that makes your .forward unique (for zealous optimizing
       sendmail that are dead wrong about optimizing!) and that can save
       you a lot of trouble if anything goes wrong! Just look at your ~/.bak!
-    * Replace /usr/local/lib/mailagent/filter with the proper filter path
+    * Replace /usr/lib/mailagent/filter with the proper filter path
       on your machine.
 
 5d] Note that on many systems, you need to ensure your .forward will be
@@ -163,8 +179,8 @@ Subject: ... 5.1 Setting Up Mailagent
 
 7c] Look out in ~/Mail/testing. You should find there the message whose Subject
     line contained the word "test". Then make sure the other message has been
-    delivered to your regular mailbox. (Since no match occurred in your rule
-    file, the mail is left in your mailbox by default).
+    delivered to the folder incoming. (Since no match occurred, the
+    mail is left in the folder specified in the default rule).
 
 7d] TROUBLESHOOTING
     * If your mail was not properly delivered, please make sure your rule file
@@ -343,6 +359,12 @@ plus flock() if asked to do so at Configure time. However, mailagent provides
 support for NFS-secure locks and also can use non-standard locking procedures,
 configurable from within ~/.mailagent (variables "nfslock" and "mboxlock").
 
+Under Debian, locking is not supported since locks require mailagent
+to be sgid mail, and given that arbitary code can be run by any user
+(see the PERL directive), this would be a major security risk).
+
+
+
 However, it cannot support locking on a rule basis (yet!). The author is
 willing to raise the priority of that item if one comes up with a legitimate
 need for that feature that could not be worked-around by a PERL escape. ;-)
diff --git a/README b/README
index de8ce2f..3e0d96e 100644
--- a/README
+++ b/README
@@ -36,11 +36,28 @@ As in lex, the filtering automaton supports the notion of modes, each
 rule belonging to a set of modes and being applied only when the current
 working mode matches one of the modes associated with the rule.
 
+NOTE
+====
+
+You must install a basic rules fle on Debian systems, namely, one that
+has minimally somthing like
+######################################################################
+### final default rule
+{ 
+        SAVE incoming
+};
+######################################################################
+This is required since Debian MDA policy requires MDA's to be sgid
+mail to lock the /var/spool/mail directory, and making mailagent sgid
+anything would be a security risk.
+
 If you do not install any filtering rules, then some default hardwired
-rules apply. Those simply leave all the messages in your mailbox, but
-process mails whose Subject line is Command (@SH hooks). You may
-override this default behavior by writing your own set of rules,
-and maybe disable this processing entirely.
+rules apply. Those simply leave all the messages in your mailbox
+(which may cause mail to be garbled on Debian systems), but process
+mails whose Subject line is Command (@SH hooks). You may override this
+default behavior by writing your own set of rules, and maybe disable
+this processing entirely.
+
 
 I have included in the subdirectory 'examples' a set of files which are
 part of my own mail environment, in the hope that they will be useful.
diff --git a/agent/examples/mchk b/agent/examples/mchk
index d76c8f2..15f96f4 100755
--- a/agent/examples/mchk
+++ b/agent/examples/mchk
@@ -2,7 +2,7 @@
 # @(#) Scans all the mailboxes to spot any new mail
 
 # Compute location of the spool mailbox
-spool=/usr/spool/mail
+spool=/var/spool/mail
 if test -d /usr/mail; then
 	spool=/usr/mail;
 fi
diff --git a/agent/examples/profile b/agent/examples/profile
index fad8b90..2deeb03 100644
--- a/agent/examples/profile
+++ b/agent/examples/profile
@@ -8,7 +8,7 @@ if test "$MAIL"; then
 elif test -d /usr/mail; then
 	mailbox=/usr/mail/ram
 else
-	mailbox=/usr/spool/mail/ram
+	mailbox=/var/spool/mail/ram
 fi
 
 # List of folders to look at for new mail
diff --git a/agent/filter/logfile.c b/agent/filter/logfile.c
index bff856b..78b4e66 100644
--- a/agent/filter/logfile.c
+++ b/agent/filter/logfile.c
@@ -47,6 +47,8 @@
 #include <stdio.h>
 #include <errno.h>
 #include <sys/types.h>
+#include <string.h>
+#include <errno.h>
 
 #ifdef I_STDLIB
 #include <stdlib.h>
@@ -105,7 +107,6 @@ public Pid_t progpid = 0;		/* Program PID */
 
 extern Time_t time();			/* Time in seconds since the Epoch */
 extern char *strsave();			/* Save string in memory */
-extern int errno;				/* System error report variable */
 
 /* VARARGS2 */
 public void add_log(level, format, arg1, arg2, arg3, arg4, arg5)
@@ -243,6 +244,7 @@ char *where;
 #if !defined(HAS_STRERROR) && defined(HAS_SYS_ERRLIST)
 	extern int sys_nerr;					/* Size of sys_errlist[] */
 	extern char *sys_errlist[];				/* Maps error code to string */
+
 #endif
 
 #ifdef HAS_STRERROR
diff --git a/agent/magent.sh b/agent/magent.sh
index c5e716a..c921a7c 100755
--- a/agent/magent.sh
+++ b/agent/magent.sh
@@ -701,10 +701,10 @@ sub mailbox_name {
 	# override value computed by Configure.
 	$maildir = $cf'maildrop if $cf'maildrop ne '';
 	# If Configure gave a valid 'maildir', use it. Otherwise compute one now.
-	unless ($maildir ne '' && -d $maildir) {
-		$maildir = "/usr/spool/mail";		# Default spooling area
-		-d $maildir || (-d "/usr/mail" && ($maildir = "/usr/mail"));
-		-d $maildir || ($maildir = $cf'home);
+	unless ($maildir ne '' && -d "$maildir") {
+		$maildir = "/var/spool/mail";		# Default spooling area
+		-d "$maildir" || ( -d "/usr/mail" && ($maildir = "/usr/mail"));
+		-d "$maildir" || ($maildir = "$cf'home");
 	}
 	local($mbox) = $cf'user;					# Default mailbox file name
 	$mbox = $cf'mailbox if $cf'mailbox ne '';	# Priority to config variable
diff --git a/agent/man/edusers.SH b/agent/man/edusers.SH
index 2e14a17..1cb21e9 100755
--- a/agent/man/edusers.SH
+++ b/agent/man/edusers.SH
@@ -18,35 +18,35 @@ echo "Extracting agent/man/edusers.$manext (with variable substitutions)"
 $rm -f edusers.$manext
 $spitshell >edusers.$manext <<!GROK!THIS!
 .TH EDUSERS $manext
-''' @(#) Manual page for mailagent's edusers command
-'''
-''' $Id: edusers.SH 1 2006-08-24 13:24:12Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-''' $Log: edusers.SH,v $
-''' Revision 3.0.1.5  1999/07/12  13:44:10  ram
-''' patch66: updated my e-mail address
-'''
-''' Revision 3.0.1.4  1996/12/24  14:08:52  ram
-''' patch45: examples are now shown in constant-width font if possible
-'''
-''' Revision 3.0.1.3  1995/08/07  16:13:00  ram
-''' patch37: updated my e-mail address
-'''
-''' Revision 3.0.1.2  1994/10/04  17:39:34  ram
-''' patch17: changed the .TH line to point at the command's name
-'''
-''' Revision 3.0.1.1  1994/09/22  13:53:06  ram
-''' patch12: created
-'''
-''' 
+.\" @(#) Manual page for mailagent's edusers command
+.\"
+.\" $Id: edusers.SH 1 2006-08-24 13:24:12Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"  
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\" $Log: edusers.SH,v $
+.\" Revision 3.0.1.5  1999/07/12  13:44:10  ram
+.\" patch66: updated my e-mail address
+.\"
+.\" Revision 3.0.1.4  1996/12/24  14:08:52  ram
+.\" patch45: examples are now shown in constant-width font if possible
+.\"
+.\" Revision 3.0.1.3  1995/08/07  16:13:00  ram
+.\" patch37: updated my e-mail address
+.\"
+.\" Revision 3.0.1.2  1994/10/04  17:39:34  ram
+.\" patch17: changed the .TH line to point at the command's name
+.\"
+.\" Revision 3.0.1.1  1994/09/22  13:53:06  ram
+.\" patch12: created
+.\"
+.\" 
 .de Ex		\" Start of Example
 .sp
 .in +5
diff --git a/agent/man/mailagent.SH b/agent/man/mailagent.SH
index 334310c..56ee863 100755
--- a/agent/man/mailagent.SH
+++ b/agent/man/mailagent.SH
@@ -18,139 +18,139 @@ echo "Extracting agent/man/mailagent.$manext (with variable substitutions)"
 $rm -f mailagent.$manext
 $spitshell >mailagent.$manext <<!GROK!THIS!
 .TH MAILAGENT $manext "Version $VERSION-$REVISION"
-''' @(#) Manual page for mailagent's filter -- (c) ram February 1991
-'''
-''' $Id: mailagent.SH 75 2011-12-23 10:18:37Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-''' $Log: mailagent.SH,v $
-''' Revision 3.0.1.26  2001/03/17 18:08:30  ram
-''' patch72: documented new config vars: domain and hidenet
-''' patch72: various fixes from bug reports on Debian
-'''
-''' Revision 3.0.1.25  2001/03/13 13:10:47  ram
-''' patch71: documented SUBST/TR on header fields
-'''
-''' Revision 3.0.1.24  2001/01/10 16:51:38  ram
-''' patch69: changed semantics of "tome"
-''' patch69: updated POST to current practices
-''' patch69: documented biffing macros for news article
-'''
-''' Revision 3.0.1.23  1999/07/12  13:46:20  ram
-''' patch66: variables are now propagated back and forth through APPLY
-''' patch66: updated my e-mail address
-'''
-''' Revision 3.0.1.22  1999/01/13  18:10:52  ram
-''' patch64: added %Y macro for 4-digit year output
-''' patch64: agent.wait file moved from Queue to Spool
-'''
-''' Revision 3.0.1.21  1998/07/28  16:59:10  ram
-''' patch62: documented new "servshell" variable
-'''
-''' Revision 3.0.1.20  1998/03/31  14:40:01  ram
-''' patch59: added "vacfixed" and "tofake" configuration parameters
-''' patch59: new ON command to issue commands on certain days only
-''' patch59: the SERVER "set" command can now list defined variables
-''' patch59: added an example of alternate VACATION message selection
-'''
-''' Revision 3.0.1.19  1997/09/15  15:07:18  ram
-''' patch57: new -t and -f options for BEGIN and NOP
-''' patch57: all command options should now be output in bold
-''' patch57: documented _CALLOUT_ working state for callout queue
-'''
-''' Revision 3.0.1.18  1997/02/20  11:40:44  ram
-''' patch55: documents new execsafe, groupsafe and lockwarn variables
-''' patch55: the C filter can now redirect output via -o
-''' patch55: made it explicit that /pattern/i is legal
-'''
-''' Revision 3.0.1.17  1997/01/07  18:29:18  ram
-''' patch52: documented new execsafe configuration variable
-'''
-''' Revision 3.0.1.16  1996/12/24  14:12:58  ram
-''' patch45: examples are now shown in constant-width font if possible
-''' patch45: big emphasis about security issues, for RUN commands & filter
-''' patch45: new Relayed: pseudo header computations
-''' patch45: documented the %-H biffing macro with more details
-''' patch45: updated my e-mail address
-'''
-''' Revision 3.0.1.15  1995/09/15  13:56:30  ram
-''' patch43: folder compression can now deal with various compressors
-''' patch43: added locksafe, compspecs and comptag config variables
-''' patch43: many typo fixes
-'''
-''' Revision 3.0.1.14  1995/08/31  16:27:44  ram
-''' patch42: escaped various dollars to avoid shell substitution, grrr...
-'''
-''' Revision 3.0.1.13  1995/08/07  16:14:23  ram
-''' patch37: new biffing features and configuration variables
-''' patch37: new BIFF filtering command to dynamically configure biffing
-'''
-''' Revision 3.0.1.12  1995/03/21  12:56:05  ram
-''' patch35: sample vacation message now contains a Precedence: header
-'''
-''' Revision 3.0.1.11  1995/02/16  14:28:45  ram
-''' patch32: documents new -I switch and new fromfake config variable
-''' patch32: random cleanup, mainly suppressing spurious "the" articles
-'''
-''' Revision 3.0.1.10  1995/01/25  15:17:42  ram
-''' patch27: new option letter 't' for mailagent -s
-''' patch27: new commands BEEP and PROTECT
-''' patch27: new macro %a for biff messages
-'''
-''' Revision 3.0.1.9  1995/01/03  18:01:53  ram
-''' patch24: new -u option for ANNOTATE documented
-''' patch24: fixed example on the shell server command (power checking)
-''' patch24: removed quotes for SERVER -d to accommodate new option parsing
-''' patch24: added a -l switch to VACATION and extended its arguments
-''' patch24: new section documenting Rule Environment variables
-'''
-''' Revision 3.0.1.8  1994/10/29  17:41:41  ram
-''' patch20: documents the six new config variables for biffing
-''' patch20: new section dedicated to built-in mail biffing
-'''
-''' Revision 3.0.1.7  1994/10/10  10:23:36  ram
-''' patch19: typo fix
-'''
-''' Revision 3.0.1.6  1994/10/04  17:41:47  ram
-''' patch17: documents new email and mboxlock config parameters
-''' patch17: documents ~/agent.trace file and callout queue file name
-''' patch17: new %e macro available to get user's e-mail address
-''' patch17: mentions that the msgpath variable is read-only
-'''
-''' Revision 3.0.1.5  1994/09/22  13:57:09  ram
-''' patch12: documents new config parameters callout and linkdirs
-''' patch12: new filtering actions AFTER and DO
-''' patch12: variable msgpath is now defined within a PERL escape
-''' patch12: mention that PERL escape variables are available to new commands
-'''
-''' Revision 3.0.1.4  1994/07/01  14:56:20  ram
-''' patch8: documents new eleven configuration variables
-''' patch8: sub-section on timeouts has been expanded
-''' patch8: emphasize .forward optimization danger with sendmail
-''' patch8: new UMASK command
-'''
-''' Revision 3.0.1.3  1994/04/25  15:15:56  ram
-''' patch7: documented new 'fromesc' config variable
-''' patch7: forgot to insert the new -F option in the synopsis line
-'''
-''' Revision 3.0.1.2  1994/01/26  09:29:24  ram
-''' patch5: documents new tag feature for UNIQUE and RECORD
-''' patch5: documents new -F option
-''' patch5: random typo fixes
-'''
-''' Revision 3.0.1.1  1993/12/15  09:03:44  ram
-''' patch3: typo and minor fixes
-'''
-''' Revision 3.0  1993/11/29  13:48:27  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
+.\" @(#) Manual page for mailagent's filter -- (c) ram February 1991
+.\"
+.\" $Id: mailagent.SH 75 2011-12-23 10:18:37Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\" $Log: mailagent.SH,v $
+.\" Revision 3.0.1.26  2001/03/17 18:08:30  ram
+.\" patch72: documented new config vars: domain and hidenet
+.\" patch72: various fixes from bug reports on Debian
+.\"
+.\" Revision 3.0.1.25  2001/03/13 13:10:47  ram
+.\" patch71: documented SUBST/TR on header fields
+.\"
+.\" Revision 3.0.1.24  2001/01/10 16:51:38  ram
+.\" patch69: changed semantics of "tome"
+.\" patch69: updated POST to current practices
+.\" patch69: documented biffing macros for news article
+.\"
+.\" Revision 3.0.1.23  1999/07/12  13:46:20  ram
+.\" patch66: variables are now propagated back and forth through APPLY
+.\" patch66: updated my e-mail address
+.\"
+.\" Revision 3.0.1.22  1999/01/13  18:10:52  ram
+.\" patch64: added %Y macro for 4-digit year output
+.\" patch64: agent.wait file moved from Queue to Spool
+.\"
+.\" Revision 3.0.1.21  1998/07/28  16:59:10  ram
+.\" patch62: documented new "servshell" variable
+.\"
+.\" Revision 3.0.1.20  1998/03/31  14:40:01  ram
+.\" patch59: added "vacfixed" and "tofake" configuration parameters
+.\" patch59: new ON command to issue commands on certain days only
+.\" patch59: the SERVER "set" command can now list defined variables
+.\" patch59: added an example of alternate VACATION message selection
+.\"
+.\" Revision 3.0.1.19  1997/09/15  15:07:18  ram
+.\" patch57: new -t and -f options for BEGIN and NOP
+.\" patch57: all command options should now be output in bold
+.\" patch57: documented _CALLOUT_ working state for callout queue
+.\"
+.\" Revision 3.0.1.18  1997/02/20  11:40:44  ram
+.\" patch55: documents new execsafe, groupsafe and lockwarn variables
+.\" patch55: the C filter can now redirect output via -o
+.\" patch55: made it explicit that /pattern/i is legal
+.\"
+.\" Revision 3.0.1.17  1997/01/07  18:29:18  ram
+.\" patch52: documented new execsafe configuration variable
+.\"
+.\" Revision 3.0.1.16  1996/12/24  14:12:58  ram
+.\" patch45: examples are now shown in constant-width font if possible
+.\" patch45: big emphasis about security issues, for RUN commands & filter
+.\" patch45: new Relayed: pseudo header computations
+.\" patch45: documented the %-H biffing macro with more details
+.\" patch45: updated my e-mail address
+.\"
+.\" Revision 3.0.1.15  1995/09/15  13:56:30  ram
+.\" patch43: folder compression can now deal with various compressors
+.\" patch43: added locksafe, compspecs and comptag config variables
+.\" patch43: many typo fixes
+.\"
+.\" Revision 3.0.1.14  1995/08/31  16:27:44  ram
+.\" patch42: escaped various dollars to avoid shell substitution, grrr...
+.\"
+.\" Revision 3.0.1.13  1995/08/07  16:14:23  ram
+.\" patch37: new biffing features and configuration variables
+.\" patch37: new BIFF filtering command to dynamically configure biffing
+.\"
+.\" Revision 3.0.1.12  1995/03/21  12:56:05  ram
+.\" patch35: sample vacation message now contains a Precedence: header
+.\"
+.\" Revision 3.0.1.11  1995/02/16  14:28:45  ram
+.\" patch32: documents new -I switch and new fromfake config variable
+.\" patch32: random cleanup, mainly suppressing spurious "the" articles
+.\"
+.\" Revision 3.0.1.10  1995/01/25  15:17:42  ram
+.\" patch27: new option letter 't' for mailagent -s
+.\" patch27: new commands BEEP and PROTECT
+.\" patch27: new macro %a for biff messages
+.\"
+.\" Revision 3.0.1.9  1995/01/03  18:01:53  ram
+.\" patch24: new -u option for ANNOTATE documented
+.\" patch24: fixed example on the shell server command (power checking)
+.\" patch24: removed quotes for SERVER -d to accommodate new option parsing
+.\" patch24: added a -l switch to VACATION and extended its arguments
+.\" patch24: new section documenting Rule Environment variables
+.\"
+.\" Revision 3.0.1.8  1994/10/29  17:41:41  ram
+.\" patch20: documents the six new config variables for biffing
+.\" patch20: new section dedicated to built-in mail biffing
+.\"
+.\" Revision 3.0.1.7  1994/10/10  10:23:36  ram
+.\" patch19: typo fix
+.\"
+.\" Revision 3.0.1.6  1994/10/04  17:41:47  ram
+.\" patch17: documents new email and mboxlock config parameters
+.\" patch17: documents ~/agent.trace file and callout queue file name
+.\" patch17: new %e macro available to get user's e-mail address
+.\" patch17: mentions that the msgpath variable is read-only
+.\"
+.\" Revision 3.0.1.5  1994/09/22  13:57:09  ram
+.\" patch12: documents new config parameters callout and linkdirs
+.\" patch12: new filtering actions AFTER and DO
+.\" patch12: variable msgpath is now defined within a PERL escape
+.\" patch12: mention that PERL escape variables are available to new commands
+.\"
+.\" Revision 3.0.1.4  1994/07/01  14:56:20  ram
+.\" patch8: documents new eleven configuration variables
+.\" patch8: sub-section on timeouts has been expanded
+.\" patch8: emphasize .forward optimization danger with sendmail
+.\" patch8: new UMASK command
+.\"
+.\" Revision 3.0.1.3  1994/04/25  15:15:56  ram
+.\" patch7: documented new 'fromesc' config variable
+.\" patch7: forgot to insert the new -F option in the synopsis line
+.\"
+.\" Revision 3.0.1.2  1994/01/26  09:29:24  ram
+.\" patch5: documents new tag feature for UNIQUE and RECORD
+.\" patch5: documents new -F option
+.\" patch5: random typo fixes
+.\"
+.\" Revision 3.0.1.1  1993/12/15  09:03:44  ram
+.\" patch3: typo and minor fixes
+.\"
+.\" Revision 3.0  1993/11/29  13:48:27  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
 .de Ex		\" Start of Example
 .sp
 .in +5
@@ -186,7 +186,7 @@ There is a set of options which may be used when you invoke
 \fImailagent\fR yourself. Please refer to the \fBOPTIONS\fR section for
 a complete description. You may use the \fB\-h\fR option to get a cryptic
 usage reminder.
-'''
+.\"
 .SS Product Overview
 .PP
 .I Mailagent
@@ -194,7 +194,9 @@ has actually four distinct set of features, which can be used simultaneously
 or one at a time. This involves:
 .IP \(bu 5
 An @SH command processor, to remain compatible with the first implementation.
-In this simplest usage, all the mail messages are left in your mailbox,
+In this simplest usage, all the mail messages are left in your mailbox
+(or the catch all folder required on Debian systems: Please see
+.B /usr/share/doc/mailagent/SECURITY for details),
 with special processing raised on messages whose subject is \fICommand\fR.
 Please refer to the section entitled \fBUSING THE DEFAULT RULES\fR if you
 wish to use this feature.
@@ -223,7 +225,7 @@ It is possible to extend the mailagent filtering commands by implementing
 them in \fIperl\fR and then having them automagically loaded when used. Those
 extended commands will behave exactly like built in ones, as documented
 in the \fBEXTENDING FILTERING COMMANDS\fR section.
-'''
+.\"
 .SS Learning From Examples
 .PP
 It is quite possible that you will find this manual page too complex for you.
@@ -232,9 +234,9 @@ material. If you wish, you may start by looking at the examples held in the
 distribution source tree under \fIagent/examples\fR. This directory contains
 two examples of rule files (look at the README file first) and are verbosely
 commented.
-'''
-''' G e t t i n g   S t a r t e d
-'''
+.\"
+.\" G e t t i n g   S t a r t e d
+.\"
 .SH "GETTING STARTED"
 .PP
 First, you need to install a minimum configuration and see how it works. It
@@ -243,7 +245,7 @@ not work as advertised...
 .PP
 To start the installation, you have to set up a \fI~/.mailagent\fR file which is
 the main configuration file, and choose the right \fIfilter\fR program.
-'''
+.\"
 .SS "Choosing The Filter Program"
 .PP
 The distribution comes with two filter programs. One written in shell and one
@@ -281,7 +283,7 @@ As of version 3.0 PL44, I advise you to prefer the C version if you are
 concerned about security. If you are in a position where multiple architectures
 can process your \fI.forward\fR, then a shell wrapper selecting the proper
 executable based on the architecture will be required.
-'''
+.\"
 .SS "Configuring Mailagent"
 .PP
 If \fImailagent\fR is in your path, you may automatically configure a default
@@ -535,7 +537,11 @@ flock()-style locking succeeded). This variable controls how safe you want
 to be. Set it to OFF to let mailagent continue its mailbox delivery even
 though no locking was done, to ON if you want strict locking, to PARTIAL if
 you can live with partial locking. Messages not saved in a folder are
-dumped to an emergency mailbox. (optional, defaults to ON).
+dumped to an emergency mailbox. (optional, defaults to ON). On Debian
+systems, since \fImailagent\fR can not grab locks,it should always be
+left ON, or else mail garbling may occur. See
+.I /usr/share/doc/mailagent/SECURITY
+for details.
 .TP
 .I lockwarn
 This variable controls the time after which \fImailagent\fR should
@@ -597,7 +603,9 @@ Available macros are:
 .Ef
 Common locking formats are "%f.lock" and "%D/.%F.lock". Of course, to be able
 to use this feature, mailagent must not have been configured to use
-flock()-style locking only. (optional, defaults to: %f.lock).
+flock()-style locking only. (optional, defaults to: %f.lock). This has
+no effect on Debian systems, since \fImailagent\fR can not get a lock
+anyway, since it is not sgid mail.
 .TP
 .I mhprofile
 The name of the MH profile to be used. This is needed only when attempting
@@ -606,7 +614,8 @@ value \fI~/.mh_profile\fR is used.
 .TP
 .I mmdf
 Set this to ON if you wish to be able to save mail in MMDF-style mailboxes.
-(suggested: OFF, unless you use MMDF or MH).
+(suggested: OFF, unless you use MMDF or MH). This is invalid on a
+Debian system.
 .TP
 .I mmdfbox
 The value of this variable only matters when \fImmdf\fR is on. If set to ON,
@@ -717,8 +726,19 @@ Set this to ON to enable macro substitutions in rule patterns.
 (optional, defaults to: OFF).
 .TP
 .I rules
-The name of the file holding the filtering rules (optional,
-suggested: ~/.rules).
+The name of the file holding the filtering rules (optional on non
+Debian systems, suggested: ~/.rules). On Debian systems, one must have
+a minimal rules file to prevent \fImailagent\fR from trying to put
+messages into
+.I /var/spool/mail/$USER,
+since mailagent can't lock that directory to prevent mail from being
+garbled. This is because Debian policy requires all entities
+attempting locks on that directory to be
+.I sgid mail,
+and making \fImailagent\fR sgid anything would be a security
+loophole.
+    { SAVE incoming };
+ is the suggested minimal rules file.
 .TP
 .I runmax
 Timeout for RUN commands and friends. (optional, defaults to: 3600).
@@ -827,7 +847,7 @@ force one unique message (optional, defaults to: OFF).
 The minimum time elapsed between two vacation messages to a given address
 (suggested: 1d).
 .PD
-'''
+.\"
 .SS "Available Logging Levels"
 .PP
 The following log levels can be used while running mailagent:
@@ -846,7 +866,7 @@ The following log levels can be used while running mailagent:
 19	Verbose
 20	Lot more verbose
 .Ef
-'''
+.\"
 .SS "Plugging Mailagent"
 .PP
 Once you have configured mailagent in a \fI~/.mailagent\fR (where \fI~\fR
@@ -861,6 +881,8 @@ unusual messages to \fI~/.bak\fR. A sample filter shell script may be found in
 \fILib/mailagent\fR, as well as a C filter program. On some systems, it may
 be necessary to move the '|' character before the leading quote, but don't
 try this unless you have no other choice (i.e. only as a last resort).
+Also, apparently \fBExim\fR takes exeption to the exec, and even
+perhaps to the redirection -- which would be a pity.
 .PP
 It is \fIvery\fR important to redirect error messages to some file within
 your home directory. For one thing, that will get you out of trouble if
@@ -886,17 +908,23 @@ security reasons).
 Note that the \fI.forward\fR file only pipes the mail to the \fIfilter\fR
 program and does not leave any copy in the mailbox. It is up to you to decide
 in the rule file whether you want to trash the mail away or leave it in the
-mailbox. If you do not have a rule file (i.e. you left a blank entry in your
-\fI~/.mailagent\fR, or you named a non-existent file, or your file is simply
-empty), don't worry: the default action is to leave the mail in the mailbox.
-'''
+mailbox.(Note that on Debian systems \fImailagent\fR can not lock the
+spool directory, and letting it leave mail in mailbox may cause it to
+get garbled). If you do not have a rule file (i.e. you left a blank
+entry in your \fI~/.mailagent\fR, or you named a non-existent file, or
+your file is simply empty),  the default action is to leave the mail
+in the mailbox, which is not a good idea for Debian machines. Please
+onstall a minimal rules file in any case,
+ { SAVE incoming };
+ is the suggested minimal rules file.
+.\"
 .SS "Allowed Commands"
 .PP
 The allowed command file (as specified by the \fIcomfile\fR variable in
 your \fI~/.mailagent\fR) contains all the recognized and allowed commands.
 The file \fIcommands\fR held in directory \fILib/mailagent\fR should be
 copied as-is into your Spool directory.
-'''
+.\"
 .SS "Testing Your Installation"
 .PP
 Now, assuming you have set a proper \fI~/.mailagent\fR file and edited the
@@ -906,8 +934,9 @@ installation. Make sure your \fI.forward\fR is world readable and that the
 \fIfilter\fR world readable).
 Set a log-level of 20 and disable vacation mode (the \fIvacation\fR entry in the
 \fI~/.mailagent\fR should be OFF). Set the name of the rule file to
-an empty file (or a non-existing file for that matter).
-You are ready to proceed...
+an file containing a catch-all rule:
+     { SAVE incoming };
+ You are ready to proceed...
 .PP
 Send yourself a mail and give mailagent time to process your mail. The
 subject of the message should be 'test' (in fact, anything but 'Command').
@@ -923,11 +952,13 @@ analyzing mail
 in mode 'INITIAL' for ALL
 selector 'All' on '<1,->', pattern '/^Subject: [Cc]ommand/'
 matching '/^Subject: [Cc]ommand/' on 'All' (<1,->) was false
-NOTICE no match, leaving in mailbox
+selector 'All'  on '<1,->'
+matching . on 'All' (<1,->) was true
+saving in folder incoming
 XEQ (LEAVE)
 starting LEAVE
-starting SAVE /usr/spool/mail/ram
-LEFT [qm7831] in mailbox
+starting SAVE /home/ram/mail/incoming
+SAVED [qm7831] in folder incoming
 FILTERED [qm7831] from ram (Raphael Manfredi)
 mailagent continues
 mailagent exits
@@ -994,9 +1025,9 @@ for local processing, to avoid an excessive workload on the \fImailhost\fR
 machine, especially if it is a dedicated NFS server. If you are a system
 administrator installing \fImailagent\fR and expect many people to use it,
 keep this in mind.
-'''
-''' O p t i o n s
-'''
+.\"
+.\" O p t i o n s
+.\"
 .SH OPTIONS
 There is a limited set of options which may be used when calling the
 mailagent directly. Only one special option at a time may be specified.
@@ -1122,12 +1153,17 @@ program waits for a mail on its standard input. If an argument is provided, it
 is the name of a file holding one mail to be processed. This is the normal
 calling procedure from the filter, the argument being the location of the
 queued mail.
-'''
-''' D e f a u l t   R u l e s
-'''
+.\"
+.\" D e f a u l t   R u l e s
+.\"
 .SH "USING THE DEFAULT RULES"
-If you do not want to use the filtering feature of mailagent, then the
-default built-in rules will be used. Those are really simple: all the mails
+If you do not want to use the filtering feature of mailagent,
+.B (NOTE:
+This may cause mail to be garbled on Debian systems, since
+\fImailagent\fR can not lock the spol directory under Debian policy
+restrictions)
+then the default built-in rules will be used.  Those are really
+simple: all the mails
 are left in your mailbox and mails with a line "Subject: Command" anywhere in
 the message will be processed. Commands are looked for on lines starting with
 "@SH". The remaining of the line is then given to a shell for execution.
@@ -1141,7 +1177,7 @@ should return a zero exit status).
 .PP
 If you do not want to use the default rules, you may skip the remaining of this
 section.
-'''
+.\"
 .SS "Configuring Help"
 .PP
 The help text mailagent will send to people must be copied from
@@ -1158,7 +1194,7 @@ files back (parameter \fImaxsize\fR in your \fI~/.mailagent\fR file\fR).
 .PP
 You may use the default help file or design one that will give even more
 details to the poor user.
-'''
+.\"
 .SS "Distribution Files"
 .PP
 The two files \fIproglist\fR and \fIdistribs\fR held in \fILib/mailagent\fR
@@ -1210,7 +1246,7 @@ directory.
 .PP
 You may include comments in both files: all lines starting with a leading
 # will be ignored.
-'''
+.\"
 .SS "Testing Your Mail Agent"
 .PP
 It is now time to make sure your mailagent works. Send yourself the following
@@ -1245,9 +1281,9 @@ your mailagent's configuration. Retry with a log level set to 20 and look
 at the issued log messages in your Log directory. Make sure that the file
 listed in the \fIplsave\fR entry of your \fI~/.mailagent\fR is correctly
 updated after a \fImaillist\fR has been run.
-'''
-''' F i l t e r i n g   R u l e s
-'''
+.\"
+.\" F i l t e r i n g   R u l e s
+.\"
 .SH "USING THE FILTER"
 The \fImailagent\fR can also be used as a filter: mail is parsed and some
 actions are taken based on simple \fIlex\fR-like rules. Actions range from
@@ -1258,6 +1294,7 @@ rule file:
 From: root { FORWARD postmaster };
 To: gue@eiffel.fr { POST mail.gue };
 Subject: /metaconfig/ { SAVE dist };
+{ SAVE incoming };
 .Ef
 There are three distinct rules. Rules are applied in sequence, until one
 matches (so the order is important). Any mail coming from \fIroot\fR will be
@@ -1265,8 +1302,8 @@ forwarded to user \fIpostmaster\fR. A mail addressed to \fIgue@eiffel.fr\fR is
 a mail coming from a mailing list. The mail is posted on a local newsgroup
 \fImail.gue\fR. Mails whose subject contains the word "metaconfig" will be
 saved in a folder \fIdist\fR for delayed reading and will not appear in the
-main mailbox. If no rule matched, the mail is left in the mailbox.
-'''
+main mailbox. If no rule matched, the mail is left in the folder incoming.
+.\"
 .SS "Rule File Syntax"
 .PP
 Here is a non-formal description of the rule file. Parsing of the file is done
@@ -1453,7 +1490,7 @@ From:
 Note the use of the file inclusion: all the users listed in file \fIusers\fR
 will have their mail left in the system mailbox. The usual rules apply for
 these loaded patterns.
-'''
+.\"
 .SS "Selector Combination"
 .PP
 A single rule may have a various set of selectors. For instance, in the
@@ -1550,7 +1587,7 @@ From: !ram, !root
 # Matches if \fIram\fR is listed in the \fITo\fR OR the \fICc\fR line
 To Cc: ram
 .Ef
-'''
+.\"
 .SS "Minimal Header"
 .PP
 A minimal set of selectors are guaranteed to be set, regardless of the
@@ -1599,7 +1636,7 @@ Where any reply should be sent. If no \fIReply-To:\fR field is present, then
 the \fIReturn-Path\fR is used (with <> stripped out), or the \fIFrom:\fR
 line is parsed to extract the e-mail address of the author.
 .PD
-'''
+.\"
 .SS "Variables"
 .PP
 The mailagent supports user-defined variables, which are globals. They are
@@ -1664,7 +1701,7 @@ There is currently no way for erasing a variable from the database. But if
 you do not use the variable any more, it will be removed when its age
 becomes greater than the maximum age specified by the \fIagemax\fR configuration
 variable.
-'''
+.\"
 .SS "Regular Expressions"
 .PP
 All the regular expressions follow the V8 syntax, as in \fIperl\fR, with all
@@ -1731,7 +1768,7 @@ then it might be worth knowing that all the set of matching selectors are
 recorded within %&, each set terminated with a ';'. If a negated selector is
 used, then %& will record all the fields which did not contain the pattern,
 assuming the selection succeeded (otherwise nothing is recorded).
-'''
+.\"
 .SS "Available Actions"
 .PP
 The following actions are available as filtering commands. Case is irrelevant
@@ -1813,7 +1850,7 @@ that started the APPLY command). You may nest them, of course.
 .TP
 ASSIGN \fIvar value\fR
 Assign the value to the user-defined variable \fIvar\fR, which may further be
-accessed as \fI%#var\fR for macro substitution or \fI#var\fR in the TR and
+accessed as \fI'%#var'\fR for macro substitution or \fI#var\fR in the TR and
 SUBST commands in place of the variable name. Note that there is no leading
 \fI#\fR in front of the variable name. The \fIvalue\fR you provide is first
 ran through \fIperl\fR to see if it contains some arithmetic operations. If the
@@ -1994,7 +2031,8 @@ expressions, and file inclusion is allowed to get headers from a file.
 .TP
 LEAVE
 Leave incoming mail in the system mailbox. This is the default action if no
-rule matched or if no saving occurred.
+rule matched or if no saving occurred. This is not recommended on
+Debian systems.
 (Fails if mail cannot be saved)
 .TP
 MACRO [\fB\-rdp\fR] \fIname\fR [= (\fIvalue\fR, \fItype\fR)]
@@ -2286,7 +2324,9 @@ error)
 .TP
 STORE \fIfolder\fR
 Save message in the specified folder and leave a copy in the system mailbox.
-The \fIfolder\fR parameter follows the same naming conventions as in SAVE.
+The \fIfolder\fR parameter follows the same naming conventions as in
+SAVE. Again, because of locking issues, leaving mail in the mailbox is
+not recommended on Debian machines.
 (Fails if message cannot be saved either in the \fIfolder\fR or in the mailbox)
 .TP
 STRIP \fIheader_fields_list\fR
@@ -2390,7 +2430,7 @@ previous one. This is useful to store output of system commands ran by
 \fIcron\fR. Don't try to use it with an MH folder or a directory folder or
 it will behave like SAVE.
 (Fails if message cannot be written)
-'''
+.\"
 .SS "Execution Status"
 .PP
 Almost all the actions modify a variable which keeps track of the execution
@@ -2411,7 +2451,7 @@ It is unfortunate that ONCE or SELECT commands cannot make the difference
 between a non-execution and a successful execution of the specified command.
 There may be a change in the way this scheme works, but it should remain
 backward compatible.
-'''
+.\"
 .SS "Perl Escape"
 .PP
 By using the PERL command, you have the ability to perform filtering and other
@@ -2586,7 +2626,7 @@ mailagent and directly calling the routines you need. If it is not documented
 in the manual page, it may be changed without notice by any further patch.
 (And this does not say that documented features may not change also... It's
 just more unlikely, and patches would clearly state that, of course.)
-'''
+.\"
 .SS "Program Environment"
 .PP
 All the programs started by mailagent via RUN and friends inherit the
@@ -2599,7 +2639,7 @@ set PATH (if you are using the shell filter).
 All the programs are executed from within the \fIhome\fR directory. This
 includes scripts started via the PERL command and mail hooks. The latter will
 be described in detail further down.
-'''
+.\"
 .SS "File inclusion"
 .PP
 Some commands like FORWARD or KEEP allow you to specify a file name between
@@ -2613,7 +2653,7 @@ will be reported.
 The file should list each parameter (be it an address, a header or a pattern)
 on a line by itself. Shell-style comments (#) are allowed within that file and
 leading white spaces are trimmed (but not trailing spaces).
-'''
+.\"
 .SS "Macros Substitutions"
 .PP
 All the commands go through a macro substitution mechanism before being
@@ -2736,7 +2776,7 @@ Year (last two digits)
 %[To]
 Value of the header field (here To:)
 .PD
-'''
+.\"
 .SS "User-defined Macros"
 .PP
 The mailagent lets you define your own macros in two ways: at the filter
@@ -2816,7 +2856,7 @@ At the filter level, the MACRO command has three options. By default,
 the command defines a new macro by using \fIpush\fR, and the other
 options each let you access one of the other interface functions.
 Note that macro definitions persist across APPLY commands.
-'''
+.\"
 .SS "User-defined Logging"
 .PP
 Most of the time when writing a new mailagent filtering command or an
@@ -2853,7 +2893,7 @@ option will not be honored in that case. This may or may not be useful to you.
 If you call \fI&main'usr_log\fR with a non-existent logfile name, logging
 is redirected to the default system-wide logfile defined in your
 \fI~/.mailagent\fR.
-'''
+.\"
 .SS "Dynamically Loading New Code"
 .PP
 In you perl routines (user-defined commands, perl hooks, etc...), you may
@@ -2889,7 +2929,7 @@ The routine returns \fIundef\fR if the file cannot be loaded
 (non-existent file, most probably), \fB0\fR if the file was loaded but
 contained a syntax error or did not define the specified function, and \fB1\fR
 for success.
-'''
+.\"
 .SS "Using Once Commands"
 .PP
 The ONCE constructs lets you specify a given command to be run once every
@@ -2923,7 +2963,7 @@ but that would be likely to be less efficient, as the first hashing would be
 done on a fixed word, hence all the timestamps would be located in the file
 \fIHash/m/e\fR (where \fIHash\fR is the name of your hashing directory, which
 is the \fIhash\fR parameter in the configuration file).
-'''
+.\"
 .SS "Using Tags in Record and Unique"
 .PP
 Both the RECORD and UNIQUE commands let you specify a comma-separated tag list
@@ -2973,7 +3013,7 @@ To Cc: agent-users {
 If you have some rule using UNIQUE without any
 tags, it will match when at least one instance of the message has been
 recorded, no matter what tag (if any at all) was used in the first place.
-'''
+.\"
 .SS "Specifying A Period"
 .PP
 The period parameter of the ONCE commands or the \fIvacperiod\fR parameter
@@ -3006,7 +3046,7 @@ year (365 days)
 All the periods are converted internally in seconds, although you do not
 really care... Examples of valid periods range from "1m" to "136y" on a 32 bits
 machine (why ?).
-'''
+.\"
 .SS "Timeouts"
 .PP
 In order to avoid having a \fImailagent\fR waiting for a command forever, a
@@ -3045,7 +3085,7 @@ lock. If you want a secure locking policy, make sure
 times
 .I lockdelay
 is greater than \fIlockhold\fR, that parameter being "large" enough.
-'''
+.\"
 .SS "Avoiding Loops"
 .PP
 The \fImailagent\fR leaves an "X-Filter:" header on each filtered message,
@@ -3071,7 +3111,7 @@ do
 add that header line.  You can add one via ANNOTATE if you wish to prevent
 loops, in case the program to which you are feeding the message might
 return it to you in some strange way.
-'''
+.\"
 .SS "Message Files"
 .PP
 The text of the message to be sent back (for MESSAGE or NOTIFY) is read from
@@ -3284,13 +3324,13 @@ concerned).
 For those hooks which are finally ran by perl, the special @INC array has
 mailagent's own private library path prepended to it, so that \fIrequire\fR
 first looks in this place.
-'''
-''' F o l d e r s
-'''
+.\"
+.\" F o l d e r s
+.\"
 .SH "FOLDERS"
 A folder is a file or a directory which can be the target of a delivery by the
 mailagent, that is to say the argument of SAVE-like commands.
-'''
+.\"
 .SS "Folder Format"
 .PP
 By default, mails are written into folders according to the standard UNIX-style
@@ -3326,7 +3366,7 @@ line is used to specify the base name of the message, then a number is
 appended to give the name of the message file to use. That is, if there is
 no such file, the folder will look like an MH one, without any MH sequence
 file though.
-'''
+.\"
 .SS "Folder Compression"
 .PP
 If you have one or more of the widely available file compression
@@ -3423,9 +3463,9 @@ made automatically to the plain file.
 .PP
 On newly created folders the \fIcomptag\fR configuration variable is
 referenced to determine the compression type to use for the folder.
-'''
-''' M a i l   B i f f i n g
-'''
+.\"
+.\" M a i l   B i f f i n g
+.\"
 .SH MAIL BIFFING
 If you are receiving and processing mail on your own machine, then you have
 access to local mail biffing where \fImailagent\fR can warn you about new
@@ -3622,9 +3662,9 @@ should be used for formatting: since you may biff to any tty you are logged
 on, that would force \fImailagent\fR to probe the tty for its column size,
 for each possible tty where output may go, and there is no reliable portable
 way of doing that. Sorry.
-'''
-''' E x t e n d i n g   F i l t e r i n g   C o m m a n d s
-'''
+.\"
+.\" E x t e n d i n g   F i l t e r i n g   C o m m a n d s
+.\"
 .SH EXTENDING FILTERING COMMANDS
 Once you've reached the \fIexpert\fR level, and provided you have a fair
 knowledge of \fIperl\fR, you may feel the need for more advanced commands
@@ -3644,7 +3684,7 @@ in an interpreted language like \fIperl\fR. This of course once you have
 convinced yourself that it is a Good Thing to customize and extend a program
 in the same language as the one used for the core, meaning usually a fairly
 low-level language with fewer user-friendly hooks.
-'''
+.\"
 .SS Overview
 .PP
 In order to implement a new command, say FOLD, you will need to do the
@@ -3674,7 +3714,7 @@ you within the \fImailhook\fR package).
 In the following sections, we're going to describe the syntax of the
 \fInewcmd\fR file, and we'll then present some low-level internal variables
 which may be used when implementing new commands.
-'''
+.\"
 .SS New Command File Format
 .PP
 The \fInewcmd\fR file consists of a series of lines, each line describing
@@ -3711,7 +3751,7 @@ FOLD  ~/mail/cmds/fold.pl  fold  no  yes
 .Ef
 to allow FOLD even in _SEEN_ state and have it executed without modifying
 the current value of the \fIlast-command-status\fR variable.
-'''
+.\"
 .SS Writing An Implementation
 .PP
 Your perl function will be loaded when needed into the special package
@@ -3755,7 +3795,7 @@ safely use \fIdie\fR or call external library routines that use \fIdie\fR.
 If you use \fIrequire\fR, be aware that mailagent is setting up a special
 \fI@INC\fR array by putting its private library path first, so you may place
 all your \fImailagent\fR-related library files in this place.
-'''
+.\"
 .SS Special Variables
 .PP
 The following special variables (some of them marked read-only, meaning you
@@ -3815,7 +3855,7 @@ the suitable X-Filter line that should be appended in \fBall\fR the mail you
 send via mailagent, in order to avoid loops. Also when you save mails
 to a folder, it's wise adding this line in case a problem arises: you may
 then identify the culprit.
-'''
+.\"
 .SS Rule Environment
 .PP
 An action might have a legitimate desire of altering the environment for
@@ -3862,7 +3902,7 @@ its processing to initialize the \fIenv\fR package, and \fI&env'cleanup\fR
 at the end before returning. Before running the actions specified on a rule
 match, \fI&apply_rules\fR calls \fI&env'restore\fR to ensure a coherent
 view of the environment while running the actions for that particular rule.
-'''
+.\"
 .SS Altering Control Flow
 .PP
 When you want to alter control flow to perform a REJECT, a RESTART or an
@@ -3886,7 +3926,7 @@ The preferred way is to invoke the \fImailhook\fR interface functions,
 \fI&mailhook'begin\fR, \fI&mailhook'reject\fR, etc..., and that will work
 even if you redefine those functions yourself. Besides, that's the only
 interface which is likely not to be changed by new versions.
-'''
+.\"
 .SS General Purpose Routines
 .PP
 The following is a list of all the general routines you may wish to call when
@@ -4015,7 +4055,7 @@ once finished, and ABORT is perl's \fIdie\fR.
 You may also use the three functions from the \fIextern\fR package which
 manipulate persistent variables (already documented in the section dealing
 with variables) as well as the user-defined macro routines.
-'''
+.\"
 .SS Example
 .PP
 Writing your own commands is not easy, since it requires some basic knowledge
@@ -4074,7 +4114,7 @@ it has been bounced. Indeed, should the SENDBACK action be the only one
 action to be run, we do not want mailagent to LEAVE the mail in the
 mailbox because it has never been saved (this default behavior being
 a precaution only -- better safe than sorry).
-'''
+.\"
 .SS Conclusion
 .PP
 If along the way you imagine some useful commands which could be made
@@ -4094,9 +4134,9 @@ code, the mailagent providing a fixed and reliable frame and the external
 program providing the service. One immediate extension would be mailing
 list handling, using this mechanism to interface with some mailing list
 management software written in perl.
-'''
-''' G e n e r i c   M a i l   S e r v e r
-'''
+.\"
+.\" G e n e r i c   M a i l   S e r v e r
+.\"
 .SH GENERIC MAIL SERVER
 .PP
 One nice thing about mailagent is that it provides you with the basic tools to
@@ -4111,7 +4151,7 @@ equivalent. There is no notion of modes, with separate command sets for
 each mode or limited name-space visibility, at least for now, so it is not
 easy (albeit possible) to implement an ftpmail server, for instance, since
 this implies the notion of mode.
-'''
+.\"
 .SS Overview
 In order to implement a mail server command (say send \fIfile\fR, which
 would send an arbitrary file from the file system in a separate mail message),
@@ -4150,7 +4190,7 @@ Start using the command... which of course is the nicest part in this scheme!
 In the following sections, we'll learn about the syntax of the \fIcomserver\fR
 file, what \fIpowers\fR are, how the session transcript is built, what the
 command environment is, etc...
-'''
+.\"
 .SS Builtin Commands Overview
 .PP
 The mail server has a limited set of builtin commands, dealing with user
@@ -4211,7 +4251,7 @@ number of command that can be issued in one single message.
 In case you think this \fImailagent\fR feature is dangerous for your account,
 do not create the \fIroot\fR and \fIsecurity\fR powers, and do not write
 any sensitive commands.
-'''
+.\"
 .SS Builtin Commands Definition
 .PP
 Now let's have a look at those builtin commands. Passwords of sensitive
@@ -4320,7 +4360,7 @@ use of that information. If no command is specified, the new identity is
 assumed until changed by another \fIuser\fR command and all the powers
 currently held by the user are released. If no \fIe-mail\fR address
 is given, the original user ID is restored.
-'''
+.\"
 .SS Command Environment
 .PP
 There are six types of commands and variables that can be specified in server
@@ -4423,7 +4463,7 @@ clearance mechanism.
 .I user
 The effective user ID, originally the same as the uid, but may be changed
 via the \fIuser\fR builtin command.
-'''
+.\"
 .SS Session Transcript
 .PP
 A session transcript is mailed back automatically to the user who requested
@@ -4446,7 +4486,7 @@ commands, authentication based on that information is really weak. A more
 "secure" authentication is provided by the server powers, which is
 password-based. Unfortunately, the clear password has to be transmitted in the
 message itself and could be eavesdropped.
-'''
+.\"
 .SS Recording New Commands and Variables
 .PP
 Server commands and variables are defined in the \fIcomserver\fR file defined
@@ -4501,7 +4541,7 @@ function to call to execute the command; if none is specified, the name of
 the command itself is called. Shell commands may use that field to supply
 additional options, which will be inserted right after the command name and
 before any other user-supplied arguments. Others should leave this alone.
-'''
+.\"
 .SS Special Command Types
 .PP
 There are currently two special command types.
@@ -4520,7 +4560,7 @@ same name as the command itself. For example, assuming a command \fIshoot\fR,
 its help file would be expected in \fIhelpdir/shoot\fR. If no file is found
 there, mailagent looks in its public library ($privlibexp) for an help file.
 Help is provided only when the help file exists and is not zero-sized.
-'''
+.\"
 .SS Creating the Root Power
 .PP
 In order to bootstrap the server, you need to create the root power. All the
@@ -4613,7 +4653,7 @@ should be prohibited and access to \fIcron\fR forbidden in order to avoid
 automatic mail processing (since it would be possible to have cron invoke
 a \fImailagent\fR process \-or any other program for that matter\- to process
 the incoming mail in a comparable way).
-'''
+.\"
 .SS Example
 .PP
 Here is an example showing the steps involved in creating a \fIshell\fR
@@ -4757,7 +4797,7 @@ End of processing (.signature)
 The first invocation of the \fIshell\fR command fails since we lack the
 \fIshell\fR power. The string "Permission denied." is echoed by the
 command itself into file descriptor #3 and makes it to the transcript.
-'''
+.\"
 .SS Conclusion
 .PP
 The generic mail server implemented in mailagent can be used to
diff --git a/agent/man/maildist.SH b/agent/man/maildist.SH
index 5883a97..3a95de3 100755
--- a/agent/man/maildist.SH
+++ b/agent/man/maildist.SH
@@ -18,28 +18,28 @@ echo "Extracting agent/man/maildist.$manext (with variable substitutions)"
 $rm -f maildist.$manext
 $spitshell >maildist.$manext <<!GROK!THIS!
 .so man$manext/mailhelp.$manext
-''' $Id: maildist.SH 1 2006-08-24 13:24:12Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-''' $Log: maildist.SH,v $
-''' Revision 3.0.1.3  1998/07/28  17:00:19  ram
-''' patch62: fixed typo on .so line
-'''
-''' Revision 3.0.1.2  1996/12/24  14:41:41  ram
-''' patch45: don't use expanded manpath in .so directives
-'''
-''' Revision 3.0.1.1  1995/08/07  15:53:13  ram
-''' patch37: use mansrcexp on the .so line to get full expanded path
-'''
-''' Revision 3.0  1993/11/29  13:48:28  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
+.\" $Id: maildist.SH 1 2006-08-24 13:24:12Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"  
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\" $Log: maildist.SH,v $
+.\" Revision 3.0.1.3  1998/07/28  17:00:19  ram
+.\" patch62: fixed typo on .so line
+.\"
+.\" Revision 3.0.1.2  1996/12/24  14:41:41  ram
+.\" patch45: don't use expanded manpath in .so directives
+.\"
+.\" Revision 3.0.1.1  1995/08/07  15:53:13  ram
+.\" patch37: use mansrcexp on the .so line to get full expanded path
+.\"
+.\" Revision 3.0  1993/11/29  13:48:28  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
 !GROK!THIS!
 chmod 444 maildist.$manext
diff --git a/agent/man/mailhelp.SH b/agent/man/mailhelp.SH
index 3f3c56e..6e61b3f 100755
--- a/agent/man/mailhelp.SH
+++ b/agent/man/mailhelp.SH
@@ -18,31 +18,31 @@ echo "Extracting agent/man/mailhelp.$manext (with variable substitutions)"
 $rm -f mailhelp.$manext
 $spitshell >mailhelp.$manext <<!GROK!THIS!
 .TH MAILHELP $manext ram
-''' @(#) Manual page for mailagent's commands -- (c) ram February 1991
-'''
-''' $Id: mailhelp.SH 75 2011-12-23 10:18:37Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-''' $Log: mailhelp.SH,v $
-''' Revision 3.0.1.3  1999/07/12  13:46:52  ram
-''' patch66: updated my e-mail address
-'''
-''' Revision 3.0.1.2  1996/12/24  14:41:59  ram
-''' patch45: documented command forwarding
-'''
-''' Revision 3.0.1.1  1995/08/07  16:14:40  ram
-''' patch37: updated my e-mail address
-'''
-''' Revision 3.0  1993/11/29  13:48:29  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
+.\" @(#) Manual page for mailagent's commands -- (c) ram February 1991
+.\"
+.\" $Id: mailhelp.SH 75 2011-12-23 10:18:37Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\" $Log: mailhelp.SH,v $
+.\" Revision 3.0.1.3  1999/07/12  13:46:52  ram
+.\" patch66: updated my e-mail address
+.\"
+.\" Revision 3.0.1.2  1996/12/24  14:41:59  ram
+.\" patch45: documented command forwarding
+.\"
+.\" Revision 3.0.1.1  1995/08/07  16:14:40  ram
+.\" patch37: updated my e-mail address
+.\"
+.\" Revision 3.0  1993/11/29  13:48:29  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
 .SH NAME
 maildist, mailhelp, maillist, mailpatch \- mailagent's commands
 .SH SYNOPSIS
diff --git a/agent/man/maillist.SH b/agent/man/maillist.SH
index e2e8f13..d11fad9 100755
--- a/agent/man/maillist.SH
+++ b/agent/man/maillist.SH
@@ -18,25 +18,25 @@ echo "Extracting agent/man/maillist.$manext (with variable substitutions)"
 $rm -f maillist.$manext
 $spitshell >maillist.$manext <<!GROK!THIS!
 .so man$manext/mailhelp.$manext
-''' $Id: maillist.SH 1 2006-08-24 13:24:12Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-''' $Log: maillist.SH,v $
-''' Revision 3.0.1.2  1996/12/24  14:43:10  ram
-''' patch45: don't use expanded manpath in .so directives
-'''
-''' Revision 3.0.1.1  1995/08/07  15:53:20  ram
-''' patch37: use mansrcexp on the .so line to get full expanded path
-'''
-''' Revision 3.0  1993/11/29  13:48:30  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
+.\" $Id: maillist.SH 1 2006-08-24 13:24:12Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"  
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\" $Log: maillist.SH,v $
+.\" Revision 3.0.1.2  1996/12/24  14:43:10  ram
+.\" patch45: don't use expanded manpath in .so directives
+.\"
+.\" Revision 3.0.1.1  1995/08/07  15:53:20  ram
+.\" patch37: use mansrcexp on the .so line to get full expanded path
+.\"
+.\" Revision 3.0  1993/11/29  13:48:30  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
 !GROK!THIS!
 chmod 444 maillist.$manext
diff --git a/agent/man/mailpatch.SH b/agent/man/mailpatch.SH
index 1c95659..4bfa9f2 100755
--- a/agent/man/mailpatch.SH
+++ b/agent/man/mailpatch.SH
@@ -18,26 +18,26 @@ echo "Extracting agent/man/mailpatch.$manext (with variable substitutions)"
 $rm -f mailpatch.$manext
 $spitshell >mailpatch.$manext <<!GROK!THIS!
 .so man$manext/mailhelp.$manext
-'''
-''' $Id: mailpatch.SH 1 2006-08-24 13:24:12Z rmanfredi $
-'''
-''' Copyright (c) 1990-2006, Raphael Manfredi
-''' 
-''' You may redistribute only under the terms of the Artistic License,
-''' as specified in the README file that comes with the distribution.
-''' You may reuse parts of this distribution only within the terms of
-''' that same Artistic License; a copy of which may be found at the root
-''' of the source tree for mailagent 3.0.
-'''
-''' $Log: mailpatch.SH,v $
-''' Revision 3.0.1.2  1996/12/24  14:43:27  ram
-''' patch45: don't use expanded manpath in .so directives
-'''
-''' Revision 3.0.1.1  1995/08/07  15:53:32  ram
-''' patch37: use mansrcexp on the .so line to get full expanded path
-'''
-''' Revision 3.0  1993/11/29  13:48:30  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
+.\"
+.\" $Id: mailpatch.SH 1 2006-08-24 13:24:12Z rmanfredi $
+.\"
+.\" Copyright (c) 1990-2006, Raphael Manfredi
+.\" 
+.\" You may redistribute only under the terms of the Artistic License,
+.\" as specified in the README file that comes with the distribution.
+.\" You may reuse parts of this distribution only within the terms of
+.\" that same Artistic License; a copy of which may be found at the root
+.\" of the source tree for mailagent 3.0.
+.\"
+.\" $Log: mailpatch.SH,v $
+.\" Revision 3.0.1.2  1996/12/24  14:43:27  ram
+.\" patch45: don't use expanded manpath in .so directives
+.\"
+.\" Revision 3.0.1.1  1995/08/07  15:53:32  ram
+.\" patch37: use mansrcexp on the .so line to get full expanded path
+.\"
+.\" Revision 3.0  1993/11/29  13:48:30  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
 !GROK!THIS!
 chmod 444 mailpatch.$manext
diff --git a/agent/man/package.SH b/agent/man/package.SH
index bd444dc..d1a0040 100755
--- a/agent/man/package.SH
+++ b/agent/man/package.SH
@@ -18,38 +18,38 @@ echo "Extracting agent/man/package.$manext (with variable substitutions)"
 $rm -f package.$manext
 $spitshell >package.$manext <<!GROK!THIS!
 .TH PACKAGE $manext
-''' @(#) Manual page for mailagent's package command
-'''
-''' $Id: package.SH 1 2006-08-24 13:24:12Z rmanfredi $
-'''
-'''  Copyright (c) 1990-2006, Raphael Manfredi
-'''  
-'''  You may redistribute only under the terms of the Artistic License,
-'''  as specified in the README file that comes with the distribution.
-'''  You may reuse parts of this distribution only within the terms of
-'''  that same Artistic License; a copy of which may be found at the root
-'''  of the source tree for mailagent 3.0.
-'''
-'''  Original Author: Graham Stoney, 1993
-'''
-''' $Log: package.SH,v $
-''' Revision 3.0.1.4  1999/07/12  13:47:00  ram
-''' patch66: updated my e-mail address
-'''
-''' Revision 3.0.1.3  1996/12/24  14:44:21  ram
-''' patch45: examples are now shown in constant-width font if possible
-''' patch45: mentions command forwarding
-'''
-''' Revision 3.0.1.2  1995/08/07  16:14:46  ram
-''' patch37: updated my e-mail address
-'''
-''' Revision 3.0.1.1  1994/09/22  13:59:03  ram
-''' patch12: documents the edusers script and fuzzy address matching
-'''
-''' Revision 3.0  1993/11/29  13:48:31  ram
-''' Baseline for mailagent 3.0 netwide release.
-'''
-''' 
+.\" @(#) Manual page for mailagent's package command
+.\"
+.\" $Id: package.SH 1 2006-08-24 13:24:12Z rmanfredi $
+.\"
+.\"  Copyright (c) 1990-2006, Raphael Manfredi
+.\"  
+.\"  You may redistribute only under the terms of the Artistic License,
+.\"  as specified in the README file that comes with the distribution.
+.\"  You may reuse parts of this distribution only within the terms of
+.\"  that same Artistic License; a copy of which may be found at the root
+.\"  of the source tree for mailagent 3.0.
+.\"
+.\"  Original Author: Graham Stoney, 1993
+.\"
+.\" $Log: package.SH,v $
+.\" Revision 3.0.1.4  1999/07/12  13:47:00  ram
+.\" patch66: updated my e-mail address
+.\"
+.\" Revision 3.0.1.3  1996/12/24  14:44:21  ram
+.\" patch45: examples are now shown in constant-width font if possible
+.\" patch45: mentions command forwarding
+.\"
+.\" Revision 3.0.1.2  1995/08/07  16:14:46  ram
+.\" patch37: updated my e-mail address
+.\"
+.\" Revision 3.0.1.1  1994/09/22  13:59:03  ram
+.\" patch12: documents the edusers script and fuzzy address matching
+.\"
+.\" Revision 3.0  1993/11/29  13:48:31  ram
+.\" Baseline for mailagent 3.0 netwide release.
+.\"
+.\" 
 .de Ex		\" Start of Example
 .sp
 .in +5
diff --git a/agent/pl/mbox.pl b/agent/pl/mbox.pl
index ce33c91..26196f9 100644
--- a/agent/pl/mbox.pl
+++ b/agent/pl/mbox.pl
@@ -17,7 +17,7 @@
 ;# mailbox (i.e. those produced by standard mail utilities with a leading From
 ;# line stating sender and date) into the mailagent's queue. This will be
 ;# especially useful on those sites where users are not allowed to have a
-;# .forward file. By using the -f option on the mailbox in /usr/spool/mail,
+;# .forward file. By using the -f option on the mailbox in /var/spool/mail,
 ;# mail will be queued and filtered as if it had come from filter via .forward.
 package mbox;
 
diff --git a/agent/pl/parse.pl b/agent/pl/parse.pl
index e5529e8..93c6c74 100644
--- a/agent/pl/parse.pl
+++ b/agent/pl/parse.pl
@@ -616,6 +616,12 @@ sub relay_list {
 	local($host, $real);
 	local($islast) = 1;				# First line we see is the "last" inserted
 	local($received);				# Received line, verbatim
+        # The regexp /\.X$/i where X is any of offical top level domains at
+        # http://data.iana.org/TLD/tlds-alpha-by-domain.txt on 15 Aug 2006 plus the
+        # extra domain "private".
+        # The regexp is the translation into Perl syntax of the result of calling Emacs's `regexp-opt'
+        # on the list of acceptable TLDs.
+        local($tlds_rx) = qr'\.A(?:ERO|RPA|[C-GIL-OQ-UWXZ])|B(?:IZ|[ABD-JMNORSTVWYZ])|C(?:AT|O(?:M|OP)|[ACDF-IK-ORUVXYZ])|D[EJKMOZ]|E(?:DU|[CEGR-U])|F[IJKMOR]|G(?:OV|[ABD-ILMNP-UWY])|H[KMNRTU]|I(?:N(?:FO|T)|[DEL-OQ-T])|J(?:OBS|[EMOP])|K[EGHIMNRWYZ]|L[ABCIKR-VY]|M(?:IL|OBI|USEUM|[ACDGHK-Z])|N(?:AME|ET|[ACEFGILOPRUZ])|O(?:M|RG)|P(?:R(?:IVATE|O)|[AE-HK-NRSTWY])|QA|R[EOUW]|S[A-EG-ORTUVYZ]|T(?:RAVEL|[CDFGHJ-PRTVWZ])|U[AGKMSYZ]|V[ACEGINU]|W[FS]|Y[ETU]|Z[AMW]$'i;
 	local($i);
 	local($_);
 
diff --git a/agent/pl/utmp/utmp_ph.c b/agent/pl/utmp/utmp_ph.c
index 16499a0..98dc9bb 100644
--- a/agent/pl/utmp/utmp_ph.c
+++ b/agent/pl/utmp/utmp_ph.c
@@ -83,8 +83,8 @@ int main()
 	char pack[MAX_LEN];
 	char fields[MAX_LEN];
 	char buf[MAX_LEN];
-	int user_off = (int) utmp->ut_name;		/* Offset of ut_name[] */
-	int line_off = (int) utmp->ut_line;		/* Offset of ut_line[] */
+	size_t user_off = (size_t) utmp->ut_name;		/* Offset of ut_name[] */
+	size_t line_off = (size_t) utmp->ut_line;		/* Offset of ut_line[] */
 	int user_len = sizeof(utmp->ut_name);	/* Length of ut_name[] array */
 	int line_len = sizeof(utmp->ut_line);	/* Length of ut_line[] array */
 	int last_off = 0;						/* Last offset in pack format */
diff --git a/agent/test/TEST b/agent/test/TEST
index 41c588f..6977049 100755
--- a/agent/test/TEST
+++ b/agent/test/TEST
@@ -58,7 +58,6 @@ $how_many = 0;
 
 require './getopt.pl';
 &Getopt;
-
 $mailagent = 'mailagent';			# Default program (dataloaded version)
 $mailagent = 'magent' if $opt_n;	# Use non-dataloaded version
 $ENV{'MAILAGENT'} = $mailagent;
@@ -68,6 +67,7 @@ $ENV{'PATH'} = "$pwd/..:.:" . $ENV{'PATH'};
 -f '../filter/filter' && -x _ || die "No filter.\n";
 $> || die "Cannot run tests as super-user. [$<,$>]\n";
 
+
 &load_ok;		# Don't rerun successful tests if up to date
 
 # A level file indicates default loglvl
@@ -115,6 +115,7 @@ select(OK);
 $| = 1;		# We may safely interrupt
 select(STDOUT);
 
+
 foreach $dir (@tests) {
 	next unless -d $dir;
 	&run($dir);
diff --git a/agent/test/basic/config.t b/agent/test/basic/config.t
index 1cb87e6..145bd47 100644
--- a/agent/test/basic/config.t
+++ b/agent/test/basic/config.t
@@ -73,7 +73,7 @@ timezone : PST8PDT
 statfile : \$spool/mailagent.st
 rules    : ~/.rules
 rulecache: ~/.cache
-maildrop : $pwd			# Do not LEAVE messages in /usr/spool/mail
+maildrop : $pwd			# Do not LEAVE messages in /var/spool/mail
 mailbox  : \$user		# Use config variable, not current perl $user
 #fromesc : ON			# Backward compatibility -- should be ON when absent
 locksafe : OFF			# Don't bother with failed locks (for fsn <= 14 chars)
-- 
2.0.0.rc0

